test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java

```diff
@@ -854,6 +854,32 @@ private static Settings.Builder getExcludeSettings(int num, Settings.Builder bui
         return builder;
     }
 
+    /**
+     * Waits for the specified data stream to have the expected number of backing indices.
+     */
+    public static List<String> waitForDataStreamBackingIndices(String dataStreamName, int expectedSize) {
+        return waitForDataStreamIndices(dataStreamName, expectedSize, false);
+    }
+
+    /**
+     * Waits for the specified data stream to have the expected number of backing or failure indices.
+     */
+    public static List<String> waitForDataStreamIndices(String dataStreamName, int expectedSize, boolean failureStore) {
+        final var clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final var listener = ClusterServiceUtils.addTemporaryStateListener(clusterService, clusterState -> {
```

Comment:
David Turner suggested using a temporary cluster state listener instead of an `assertBusy` that uses the GET data streams transport action. This has improved performance as we can react to new cluster state updates immediately instead of waiting for the exponential backoff of `assertBusy`. The only "drawback" is that an `assertBusy` would implicitly cover the GET data streams transport action. I don't think we should rely on tests like these to cover that action - i.e. we should have dedicated tests that sufficiently cover the action, and I think we do.

Reply:
I would rather not add the override that returns the cluster state like this. Instead, just await for the cluster to reach a suitable state and then use the regular APIs and transport actions to query what you need. That way you get the little bit of extra coverage from calling the transport action, and you also get some confidence that the state you were waiting for doesn't immediately vanish in the next update since the transport action may see a future state.

Reply:
Hm, that reduces the performance benefit because we'll have to do another request, but only by a little bit I think. I agree that it does bring back some extra coverage. I'll revert the changes to `addTemporaryStateListener` and use the API to fetch the backing indices.

Reply:
The performance benefits are all about avoiding the `Thread.sleep` in the `assertBusy` call, oversleeping past the point where the cluster state is ready for you. The cost of the transport action should be very low.